python/perf-kernels/fused_moe/moe-gemm.py

```diff
@@ -0,0 +1,503 @@
+    config = get_config_func(M)
+    sorted_token_ids, expert_ids, num_tokens_post_padded = moe_align_block_size(topk_ids, config['BLOCK_SIZE_M'], E)
+
+    if not routed_weight:
+        return a, b, c, None, topk_ids, sorted_token_ids, expert_ids, num_tokens_post_padded, config
+
+    return a, b, c, topk_weights, topk_ids, sorted_token_ids, expert_ids, num_tokens_post_padded, config
+
+
+@pytest.mark.parametrize("M, N, K, top_k, E", [
+    (64, 14336, 4096, 2, 8),
+    (16, 14336, 1, 2, 4),
+    (1, 14336, 128, 2, 4),
+    (16, 14336, 128, 1, 4),
+    (16, 14336, 128, 1, 1),
+    (64, 7186, 128, 2, 8),
+    (64, 3584, 128, 2, 8),
+    (64, 1792, 128, 2, 8),
+    (64, 64, 128, 2, 8),
+])
+@pytest.mark.parametrize('routed_weight', [True, False])
+def test_correctness(M: int, N: int, K: int, top_k: int, E: int, routed_weight: bool, dtype=torch.float16):
+    torch.manual_seed(20)
+    a, b, c, topk_weights, topk_ids, sorted_token_ids, expert_ids, num_tokens_post_padded, config = input_helper(
+        M, N, K, top_k, E, routed_weight=routed_weight, dtype=dtype)
+
+    tri_out = moe_gemm(a, b, c, topk_weights, topk_ids, sorted_token_ids, expert_ids, num_tokens_post_padded, config)
+
+    ref_out = torch.empty_like(c)
+    # Repeat a -> (M, top_k, K)
+    a_expanded = a.unsqueeze(1).repeat(1, top_k, 1)
+    # (M, top_k, N, K)
+    b_indexed = b[topk_ids]
+    ref_out = torch.einsum("mek,menk->men", a_expanded, b_indexed)
+    if routed_weight:
+        ref_out *= topk_weights.unsqueeze(-1)
+
+    # Validate correctness
+    torch.testing.assert_close(tri_out, ref_out, atol=1e-2, rtol=1e-2)
+
+
+def get_configs():
+    configs = [
+        {"M": 64, "N": 256, "K": 128, "E": 8, "top_k": 2},
+        {"M": 64, "N": 1792, "K": 1024, "E": 8, "top_k": 2},
+        {"M": 64, "N": 7168, "K": 4096, "E": 8, "top_k": 2},
+        {"M": 128, "N": 7168, "K": 4096, "E": 8, "top_k": 2},
+        {"M": 1024, "N": 7168, "K": 4096, "E": 8, "top_k": 2},
+        {"M": 4096, "N": 7168, "K": 4096, "E": 8, "top_k": 2},
+        {"M": 64, "N": 14336, "K": 4096, "E": 8, "top_k": 2},
+        {"M": 128, "N": 14336, "K": 4096, "E": 8, "top_k": 2},
+        {"M": 256, "N": 14336, "K": 4096, "E": 8, "top_k": 2},
+        {"M": 512, "N": 14336, "K": 4096, "E": 8, "top_k": 2},
+        {"M": 1024, "N": 14336, "K": 4096, "E": 8, "top_k": 2},
+        {"M": 2048, "N": 14336, "K": 4096, "E": 8, "top_k": 2},
+        {"M": 4096, "N": 14336, "K": 4096, "E": 8, "top_k": 2},
+    ]
+    return configs
+
+
+def model_benchmark_configs(args):
+    config_file = args.model_configs
+    configs = get_model_configs(config_path=config_file, model_families=["mistral"], model=args.model)
+    fa_configs = []
+    M = args.M if args.M else 4096  # check size
+    # M, K, N, E, top_k
+
+    for model_name, config in configs.items():
+        N1 = config["intermediate_size"]
+        K1 = config["hidden_size"]
+
+        N2 = config["hidden_size"]
+        K2 = config["intermediate_size"] // 2
```

Comment:
why `// 2`?

Reply:
![image](https://github.com/user-attachments/assets/d1ae3176-85d7-4b1d-ab59-4866033dcadc)
We are benchmarking for the first and second gemm here for the mistral model

Reply:
Can you share the original doc where this table come from.